File: test-suite/python/security/constant_time_compare_buggy.py
import hmac

from flask import request, session


def webhook_signature_check(payload, signing_secret):
    expected_signature = hmac.new(signing_secret, payload, "sha256").hexdigest()
    provided_signature = request.headers["X-Signature"]
    return provided_signature == expected_signature


def api_key_check(request_api_key, stored_api_key):
    if request_api_key != stored_api_key:
        return False
    return True


def csrf_token_check(csrf_token):
    assert csrf_token == session["csrf_token"]


def password_reset_check(token, user):
    return token == user.reset_token


def webhook_digest_check(payload, provided_sig):
    digest = calculate_webhook_digest(payload)
    if digest == provided_sig:
        return "accepted"
    return "rejected"


def bearer_token_check(headers, expected_token):
    return headers["Authorization"] == expected_token


def calculate_webhook_digest(payload):
    return hmac.new(b"key", payload, "sha256").hexdigest()

File: test-suite/golang/security/constant_time_compare_buggy.go
package security

import "net/http"

type account struct {
	ResetToken string
	APIKey     string
}

func verifyWebhookSignature(r *http.Request, expectedSignature string) bool {
	providedSignature := r.Header.Get("X-Signature")
	return providedSignature == expectedSignature
}

func verifyAPIKey(requestAPIKey string, storedAPIKey string) bool {
	if requestAPIKey != storedAPIKey {
		return false
	}
	return true
}

func verifyCSRF(r *http.Request, sessionCSRFToken string) bool {
	csrfToken := r.FormValue("csrf_token")
	return csrfToken == sessionCSRFToken
}

func verifyResetToken(token string, user account) bool {
	return token == user.ResetToken
}

func verifyBearerToken(r *http.Request, expectedToken string) bool {
	authorization := r.Header.Get("Authorization")
	return authorization == expectedToken
}

func rejectWrongResetToken(token string, user account) bool {
	return token != user.ResetToken
}

func verifySignatureInline(signature string, expectedSignature string) bool {
	return signature == expectedSignature
}
